File: core/src/ru/myitschool/dcrawler/entities/skills/Play.java
package ru.myitschool.dcrawler.entities.skills;

import com.badlogic.gdx.utils.Array;
import ru.myitschool.dcrawler.entities.skills.action.Action;
import ru.myitschool.dcrawler.entities.skills.check.PlayCheck;

/**
 * Created by Voyager on 28.06.2017.
 */
public abstract class Play {

    public static final int TARGETING_CRIT_MISS = 0;
    public static final int TARGETING_MISS = 1;
    public static final int TARGETING_HIT = 2;
    public static final int TARGETING_CRIT_HIT = 3;

    private PlayCheck playCheck;

    private Skill skill;

    private Array<ru.myitschool.dcrawler.entities.skills.action.Action> actions = new Array<ru.myitschool.dcrawler.entities.skills.action.Action>();
    private Array<Play> plays = new Array<>();

    public Play(Skill skill) {
        this.skill = skill;
        this.playCheck = new PlayCheck(skill) {
            @Override
            public int check(Target target) {
                return TARGETING_MISS;
            }
        };
    }

    public abstract boolean isValidTarget(Target target);

    public int check(Target target){
        return playCheck.check(target);
    }

    public Play addAction(ru.myitschool.dcrawler.entities.skills.action.Action action){
        actions.add(action);
        return this;
    }

    public Play addPlay(Play play){
        plays.add(play);
        return this;
    }

    public void act(Target target, FloatingDamageMark mark){
        if (isValidTarget(target)) {
            int success = check(target);
            for (Action action : actions) {
                action.act(target, success, mark);
            }
            for (Play play : plays){
                play.act(target, mark);
            }
        }
    }

    public Play setPlayCheck(PlayCheck playCheck) {
        this.playCheck = playCheck;
        return this;
    }
}

File: core/src/com/wolg_vlad/dcrawler/entities/skills/Play.kt
package com.wolg_vlad.dcrawler.entities.skills

import com.badlogic.gdx.utils.Array
import com.wolg_vlad.dcrawler.entities.skills.action.Action
import com.wolg_vlad.dcrawler.entities.skills.check.PlayCheck

/**
 * Created by Voyager on 28.06.2017.
 */
abstract class Play(private val skill: Skill) {
    private var playCheck: PlayCheck
    private val actions = Array<Action>()
    private val plays = Array<Play>()
    abstract fun isValidTarget(target: Target): Boolean
    fun check(target: Target): Int {
        return playCheck.check(target)
    }

    fun addAction(action: Action): Play {
        actions.add(action)
        return this
    }

    fun addPlay(play: Play): Play {
        plays.add(play)
        return this
    }

    fun act(target: Target, mark: FloatingDamageMark) {
        if (isValidTarget(target)) {
            val success = check(target)
            for (action in actions) {
                action.act(target, success, mark)
            }
            for (play in plays) {
                play.act(target, mark)
            }
        }
    }

    fun setPlayCheck(playCheck: PlayCheck): Play {
        this.playCheck = playCheck
        return this
    }

    companion object {
        const val TARGETING_CRIT_MISS = 0
        const val TARGETING_MISS = 1
        const val TARGETING_HIT = 2
        const val TARGETING_CRIT_HIT = 3
    }

    init {
        playCheck = object : PlayCheck(skill) {
            override fun check(target: Target): Int {
                return TARGETING_MISS
            }
        }
    }
}